File: scripts/generate-sitemap.js
const fs = require('fs');
const problems = require('../data/problems.json');
const wiki = require('../data/wiki.json');

const SITEMAP_PATH = './public/sitemap.xml';
const BASE_URL = 'https://derivativecalculatorai.com';

function generateSitemap() {
  const locales = ['', 'es', 'pt'];

  // Reuse the prior lastmod for any URL that already existed, so regenerating
  // the sitemap does NOT churn every timestamp. Only genuinely new URLs (e.g.
  // freshly added tag pages) get a fresh date. This keeps commits clean and
  // avoids lastmod-only noise while still dated-stamping new entries.
  const prevLastmod = new Map();
  try {
    const prev = fs.readFileSync(SITEMAP_PATH, 'utf8');
    const re = /<loc>([^<]+)<\/loc>\s*<lastmod>([^<]+)<\/lastmod>/g;
    let m;
    while ((m = re.exec(prev)) !== null) prevLastmod.set(m[1], m[2]);
  } catch (e) {
    // No prior sitemap yet — every URL gets a fresh timestamp below.
  }
  const lm = (loc) => prevLastmod.get(loc) || new Date().toISOString();

  // Static Pages (updated with new SEO architecture)
  const staticPages = ['', 'integral', 'limit', 'matrix', 'ode', 'directory', 'wiki', 'problems', 'calculators', 'practice'];

  const staticUrls = staticPages.flatMap(page => {
    return locales.map(locale => {
      const urlPath = locale ? (page ? `${locale}/${page}` : locale) : page;
      const loc = `${BASE_URL}${urlPath ? '/' + urlPath : ''}`;
      return `
  <url>
    <loc>${loc}</loc>
    <lastmod>${lm(loc)}</lastmod>
    <changefreq>weekly</changefreq>
    <priority>${!urlPath ? '1.0' : '0.9'}</priority>
  </url>`;
    });
  });

  const wikiUrls = wiki.flatMap((topic) => {
    return locales.map(locale => {
      const urlPath = locale ? `${locale}/wiki/${topic.slug}` : `wiki/${topic.slug}`;
      const loc = `${BASE_URL}/${urlPath}`;
      return `
  <url>
    <loc>${loc}</loc>
    <lastmod>${lm(loc)}</lastmod>
    <changefreq>monthly</changefreq>
    <priority>0.7</priority>
  </url>`;
    });
  });

  const problemUrls = problems.flatMap((problem) => {
    return locales.map(locale => {
      const urlPath = locale ? `${locale}/${problem.slug}` : problem.slug;
      const loc = `${BASE_URL}/${urlPath}`;
      return `
  <url>
    <loc>${loc}</loc>
    <lastmod>${lm(loc)}</lastmod>
    <changefreq>monthly</changefreq>
    <priority>0.6</priority>
  </url>`;
    });
  });

  // Tag pages: emit ONLY real tags present in the data. This mirrors the tag
  // page's indexability rule (generateMetadata sets index,follow only when a
  // tag has >=1 problem). Empty/nonexistent tags are never added, so thin URLs
  // stay out of the index. 3 locales to match the rest of the sitemap.
  const tagSet = new Set();
  for (const p of problems) {
    if (p && p.tags) {
      for (const raw of String(p.tags).split(',')) {
        const t = raw.trim();
        if (t) tagSet.add(t);
      }
    }
  }
  const realTags = Array.from(tagSet);
  const tagUrls = realTags.flatMap((tag) => {
    return locales.map(locale => {
      const urlPath = locale ? `${locale}/problems/tag/${tag}` : `problems/tag/${tag}`;
      const loc = `${BASE_URL}/${urlPath}`;
      return `
  <url>
    <loc>${loc}</loc>
    <lastmod>${lm(loc)}</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.7</priority>
  </url>`;
    });
  });

  const urls = [...staticUrls, ...wikiUrls, ...problemUrls, ...tagUrls];

  const sitemap = `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
${urls.join('')}
</urlset>`;

  fs.writeFileSync(SITEMAP_PATH, sitemap);
  console.log(`Sitemap generated at ${SITEMAP_PATH}`);
}

generateSitemap();

File: scripts/generate_sitemap.py
import json
import datetime
import os

PROBLEMS_FILE = "data/problems.json"
SITEMAP_FILE = "public/sitemap.xml"
BASE_URL = "https://derivativecalculatorai.com"

def generate_sitemap():
    # Load problems
    with open(PROBLEMS_FILE, "r") as f:
        problems = json.load(f)

    # XML Header
    xml_content = '<?xml version="1.0" encoding="UTF-8"?>\n'
    xml_content += '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n'

    # Static Pages (Home, Tools, Directory, Wiki Index, Authority)
    static_tools = ["", "integral", "limit", "matrix", "ode", "directory", "wiki", "about", "contact"]
    today = datetime.date.today().isoformat()

    for page in static_tools:
        for loc in ["", "es", "pt"]:
            path = f"/{loc}/{page}" if loc else f"/{page}"
            url = f"{BASE_URL}{path}".rstrip("/")
            if url == BASE_URL: url = f"{BASE_URL}/" # Ensure home is trailing slashed or consistent
            
            xml_content += "  <url>\n"
            xml_content += f"    <loc>{url}</loc>\n"
            xml_content += f"    <lastmod>{today}</lastmod>\n"
            xml_content += "    <changefreq>daily</changefreq>\n"
            xml_content += "    <priority>1.0</priority>\n"
            xml_content += "  </url>\n"

    # Dynamic Problem Pages
    for problem in problems:
        # English
        url = f"{BASE_URL}/{problem['slug']}"
        xml_content += "  <url>\n"
        xml_content += f"    <loc>{url}</loc>\n"
        xml_content += f"    <lastmod>{today}</lastmod>\n"
        xml_content += "    <changefreq>weekly</changefreq>\n"
        xml_content += "    <priority>0.8</priority>\n"
        xml_content += "  </url>\n"

        # Spanish
        url_es = f"{BASE_URL}/es/{problem['slug']}"
        xml_content += "  <url>\n"
        xml_content += f"    <loc>{url_es}</loc>\n"
        xml_content += f"    <lastmod>{today}</lastmod>\n"
        xml_content += "    <changefreq>weekly</changefreq>\n"
        xml_content += "    <priority>0.8</priority>\n"
        xml_content += "  </url>\n"

        # Portuguese
        url_pt = f"{BASE_URL}/pt/{problem['slug']}"
        xml_content += "  <url>\n"
        xml_content += f"    <loc>{url_pt}</loc>\n"
        xml_content += f"    <lastmod>{today}</lastmod>\n"
        xml_content += "    <changefreq>weekly</changefreq>\n"
        xml_content += "    <priority>0.8</priority>\n"
        xml_content += "  </url>\n"

    # Wiki Articles (Synced from wiki.json)
    wiki_path = "content/wiki.json"
    if os.path.exists(wiki_path):
        with open(wiki_path, "r") as f:
            wiki_data = json.load(f)
            for article in wiki_data:
                url = f"{BASE_URL}/wiki/{article['slug']}"
                xml_content += "  <url>\n"
                xml_content += f"    <loc>{url}</loc>\n"
                xml_content += f"    <lastmod>{today}</lastmod>\n"
                xml_content += "    <changefreq>monthly</changefreq>\n"
                xml_content += "    <priority>0.7</priority>\n"
                xml_content += "  </url>\n"

    xml_content += '</urlset>'

    # Ensure public dir exists (Next.js standard)
    os.makedirs("public", exist_ok=True)

    with open(SITEMAP_FILE, "w") as f:
        f.write(xml_content)

    total_urls = (len(static_tools) * 3) + (len(problems) * 3)
    print(f"Generated sitemap.xml with {total_urls} URLs.")

if __name__ == "__main__":
    generate_sitemap()
